### PR TITLE
Add protocol_version to lb target groups

### LINKED ISSUE #15929: Support the new Application Load Balancer gRPC workload & HTTP/2 protocol versions
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS have added support for gRPC workloads with end to end HTTP/2 to the Application Load Balancer.  The `aws_lb_target_group` resource needs to be enhanced to support these additional protocol versions.

### New or Affected Resource(s)

* aws_lb_target_group

### Potential Terraform Configuration

```hcl
# gRPC
resource "aws_lb_target_group" "grpc" {
  name             = "tf-example-lb-tg"
  port             = 5500
  protocol         = "HTTPS"
  protocol_version = "gRPC"
  vpc_id           = aws_vpc.main.id
}

# http2
resource "aws_lb_target_group" "http2" {
  name             = "tf-example-lb-tg"
  port             = 443
  protocol         = "HTTPS"
  protocol_version = "HTTP2"
  vpc_id           = aws_vpc.main.id
}
```

### References

* https://aws.amazon.com/about-aws/whats-new/2020/10/application-load-balancers-enable-grpc-workloads-end-to-end-http-2-support/
* https://aws.amazon.com/blogs/aws/new-application-load-balancer-support-for-end-to-end-http-2-and-grpc/

### PR DESCRIPTION
**Pull request by fivehole68**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #15929

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Add protocol version support to aws_lb_target_group
```

Output from acceptance testing:

```
make testacc TESTARGS='-run=TestAccAWSLBTargetGroup' ACCTEST_PARALLELISM='2'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSLBTargetGroup -timeout 120m
=== RUN   TestAccAWSLBTargetGroupAttachment_basic
=== PAUSE TestAccAWSLBTargetGroupAttachment_basic
=== RUN   TestAccAWSLBTargetGroupAttachment_disappears
=== PAUSE TestAccAWSLBTargetGroupAttachment_disappears
=== RUN   TestAccAWSLBTargetGroupAttachment_BackwardsCompatibility
=== PAUSE TestAccAWSLBTargetGroupAttachment_BackwardsCompatibility
=== RUN   TestAccAWSLBTargetGroupAttachment_Port
=== PAUSE TestAccAWSLBTargetGroupAttachment_Port
=== RUN   TestAccAWSLBTargetGroupAttachment_ipAddress
=== PAUSE TestAccAWSLBTargetGroupAttachment_ipAddress
=== RUN   TestAccAWSLBTargetGroupAttachment_lambda
=== PAUSE TestAccAWSLBTargetGroupAttachment_lambda
=== RUN   TestAccAWSLBTargetGroup_basic
=== PAUSE TestAccAWSLBTargetGroup_basic
=== RUN   TestAccAWSLBTargetGroup_basicUdp
=== PAUSE TestAccAWSLBTargetGroup_basicUdp
=== RUN   TestAccAWSLBTargetGroup_withoutHealthcheck
=== PAUSE TestAccAWSLBTargetGroup_withoutHealthcheck
=== RUN   TestAccAWSLBTargetGroup_networkLB_TargetGroup
=== PAUSE TestAccAWSLBTargetGroup_networkLB_TargetGroup
=== RUN   TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol
=== PAUSE TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol
=== RUN   TestAccAWSLBTargetGroup_Protocol_Tls
=== PAUSE TestAccAWSLBTargetGroup_Protocol_Tls
=== RUN   TestAccAWSLBTargetGroup_ProtocolVersion_GRPC_HealthCheck
=== PAUSE TestAccAWSLBTargetGroup_ProtocolVersion_GRPC_HealthCheck
=== RUN   TestAccAWSLBTargetGroup_ProtocolVersion_HTTP_GRPC_Update
=== PAUSE TestAccAWSLBTargetGroup_ProtocolVersion_HTTP_GRPC_Update
=== RUN   TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy
=== PAUSE TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy
=== RUN   TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck
=== PAUSE TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck
=== RUN   TestAccAWSLBTargetGroup_BackwardsCompatibility
=== PAUSE TestAccAWSLBTargetGroup_BackwardsCompatibility
=== RUN   TestAccAWSLBTargetGroup_namePrefix
=== PAUSE TestAccAWSLBTargetGroup_namePrefix
=== RUN   TestAccAWSLBTargetGroup_generatedName
=== PAUSE TestAccAWSLBTargetGroup_generatedName
=== RUN   TestAccAWSLBTargetGroup_changeNameForceNew
=== PAUSE TestAccAWSLBTargetGroup_changeNameForceNew
=== RUN   TestAccAWSLBTargetGroup_changeProtocolForceNew
=== PAUSE TestAccAWSLBTargetGroup_changeProtocolForceNew
=== RUN   TestAccAWSLBTargetGroup_changePortForceNew
=== PAUSE TestAccAWSLBTargetGroup_changePortForceNew
=== RUN   TestAccAWSLBTargetGroup_changeVpcForceNew
=== PAUSE TestAccAWSLBTargetGroup_changeVpcForceNew
=== RUN   TestAccAWSLBTargetGroup_tags
=== PAUSE TestAccAWSLBTargetGroup_tags
=== RUN   TestAccAWSLBTargetGroup_enableHealthCheck
=== PAUSE TestAccAWSLBTargetGroup_enableHealthCheck
=== RUN   TestAccAWSLBTargetGroup_updateHealthCheck
=== PAUSE TestAccAWSLBTargetGroup_updateHealthCheck
=== RUN   TestAccAWSLBTargetGroup_updateSticknessEnabled
=== PAUSE TestAccAWSLBTargetGroup_updateSticknessEnabled
=== RUN   TestAccAWSLBTargetGroup_defaults_application
=== PAUSE TestAccAWSLBTargetGroup_defaults_application
=== RUN   TestAccAWSLBTargetGroup_defaults_network
=== PAUSE TestAccAWSLBTargetGroup_defaults_network
=== RUN   TestAccAWSLBTargetGroup_stickinessDefaultNLB
=== PAUSE TestAccAWSLBTargetGroup_stickinessDefaultNLB
=== RUN   TestAccAWSLBTargetGroup_stickinessDefaultALB
=== PAUSE TestAccAWSLBTargetGroup_stickinessDefaultALB
=== RUN   TestAccAWSLBTargetGroup_stickinessValidNLB
=== PAUSE TestAccAWSLBTargetGroup_stickinessValidNLB
=== RUN   TestAccAWSLBTargetGroup_stickinessValidALB
=== PAUSE TestAccAWSLBTargetGroup_stickinessValidALB
=== RUN   TestAccAWSLBTargetGroup_stickinessInvalidNLB
=== PAUSE TestAccAWSLBTargetGroup_stickinessInvalidNLB
=== RUN   TestAccAWSLBTargetGroup_stickinessInvalidALB
=== PAUSE TestAccAWSLBTargetGroup_stickinessInvalidALB
=== CONT  TestAccAWSLBTargetGroupAttachment_basic
=== CONT  TestAccAWSLBTargetGroup_stickinessValidALB
--- PASS: TestAccAWSLBTargetGroup_stickinessValidALB (45.40s)
=== CONT  TestAccAWSLBTargetGroup_stickinessInvalidALB
--- PASS: TestAccAWSLBTargetGroup_stickinessInvalidALB (41.07s)
=== CONT  TestAccAWSLBTargetGroup_stickinessInvalidNLB
--- PASS: TestAccAWSLBTargetGroupAttachment_basic (89.69s)
=== CONT  TestAccAWSLBTargetGroup_BackwardsCompatibility
--- PASS: TestAccAWSLBTargetGroup_BackwardsCompatibility (24.48s)
=== CONT  TestAccAWSLBTargetGroup_stickinessValidNLB
--- PASS: TestAccAWSLBTargetGroup_stickinessInvalidNLB (41.66s)
=== CONT  TestAccAWSLBTargetGroup_stickinessDefaultALB
--- PASS: TestAccAWSLBTargetGroup_stickinessDefaultALB (24.15s)
=== CONT  TestAccAWSLBTargetGroup_stickinessDefaultNLB
--- PASS: TestAccAWSLBTargetGroup_stickinessValidNLB (89.70s)
=== CONT  TestAccAWSLBTargetGroup_defaults_network
--- PASS: TestAccAWSLBTargetGroup_stickinessDefaultNLB (68.40s)
=== CONT  TestAccAWSLBTargetGroup_defaults_application
--- PASS: TestAccAWSLBTargetGroup_defaults_network (31.83s)
=== CONT  TestAccAWSLBTargetGroup_updateSticknessEnabled
--- PASS: TestAccAWSLBTargetGroup_defaults_application (24.23s)
=== CONT  TestAccAWSLBTargetGroup_updateHealthCheck
--- PASS: TestAccAWSLBTargetGroup_updateHealthCheck (47.65s)
=== CONT  TestAccAWSLBTargetGroup_enableHealthCheck
--- PASS: TestAccAWSLBTargetGroup_updateSticknessEnabled (65.86s)
=== CONT  TestAccAWSLBTargetGroup_tags
--- PASS: TestAccAWSLBTargetGroup_enableHealthCheck (31.86s)
=== CONT  TestAccAWSLBTargetGroup_changeNameForceNew
--- PASS: TestAccAWSLBTargetGroup_tags (67.55s)
=== CONT  TestAccAWSLBTargetGroup_generatedName
--- PASS: TestAccAWSLBTargetGroup_changeNameForceNew (47.15s)
=== CONT  TestAccAWSLBTargetGroup_namePrefix
--- PASS: TestAccAWSLBTargetGroup_generatedName (23.89s)
=== CONT  TestAccAWSLBTargetGroup_withoutHealthcheck
--- PASS: TestAccAWSLBTargetGroup_namePrefix (23.71s)
=== CONT  TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck
--- PASS: TestAccAWSLBTargetGroup_withoutHealthcheck (16.96s)
=== CONT  TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy
--- PASS: TestAccAWSLBTargetGroup_TCP_HTTPHealthCheck (45.09s)
=== CONT  TestAccAWSLBTargetGroup_ProtocolVersion_HTTP_GRPC_Update
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroupWithProxy (44.74s)
=== CONT  TestAccAWSLBTargetGroup_ProtocolVersion_GRPC_HealthCheck
--- PASS: TestAccAWSLBTargetGroup_ProtocolVersion_GRPC_HealthCheck (24.55s)
=== CONT  TestAccAWSLBTargetGroup_Protocol_Tls
--- PASS: TestAccAWSLBTargetGroup_ProtocolVersion_HTTP_GRPC_Update (40.97s)
=== CONT  TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tls (20.37s)
=== CONT  TestAccAWSLBTargetGroup_changeProtocolForceNew
--- PASS: TestAccAWSLBTargetGroup_Protocol_Tcp_HealthCheck_Protocol (48.13s)
=== CONT  TestAccAWSLBTargetGroup_changeVpcForceNew
--- PASS: TestAccAWSLBTargetGroup_changeProtocolForceNew (48.01s)
=== CONT  TestAccAWSLBTargetGroup_changePortForceNew
--- PASS: TestAccAWSLBTargetGroup_changeVpcForceNew (47.32s)
=== CONT  TestAccAWSLBTargetGroup_networkLB_TargetGroup
--- PASS: TestAccAWSLBTargetGroup_changePortForceNew (47.30s)
=== CONT  TestAccAWSLBTargetGroupAttachment_ipAddress
--- PASS: TestAccAWSLBTargetGroup_networkLB_TargetGroup (78.03s)
=== CONT  TestAccAWSLBTargetGroup_basic
--- PASS: TestAccAWSLBTargetGroup_basic (24.75s)
=== CONT  TestAccAWSLBTargetGroup_basicUdp
--- PASS: TestAccAWSLBTargetGroup_basicUdp (24.96s)
=== CONT  TestAccAWSLBTargetGroupAttachment_BackwardsCompatibility
--- PASS: TestAccAWSLBTargetGroupAttachment_ipAddress (187.03s)
=== CONT  TestAccAWSLBTargetGroupAttachment_lambda
--- PASS: TestAccAWSLBTargetGroupAttachment_BackwardsCompatibility (80.56s)
=== CONT  TestAccAWSLBTargetGroupAttachment_Port
--- PASS: TestAccAWSLBTargetGroupAttachment_lambda (36.34s)
=== CONT  TestAccAWSLBTargetGroupAttachment_disappears
--- PASS: TestAccAWSLBTargetGroupAttachment_disappears (90.09s)
--- PASS: TestAccAWSLBTargetGroupAttachment_Port (203.65s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	990.329s
...
```

### CODE REVIEW

**Review by oliveirafilipe — APPROVED** [review]

File: aws/resource_aws_lb_target_group_test.go
Comment (kevincantu) [inline aws/resource_aws_lb_target_group_test.go:1767]: I bet the `terrafmt` error is due to the spacing here (the `=` signs aren't all in line).

File: tools/go.mod
Comment (kevincantu) [inline tools/go.mod:10]: Whether this came in by accident or from the merge, it seems worth rebasing and taking this out.

I can help you with how to rebase if you want a second pair of eyes!

**Comment by kevincantu** [comment]
An idle thought: can you rename this pull request?  I assume the "f" stands for your name?

I am not expert in the code being modified, but this handles _a lot_ more of the `protocol` / `protocol_version` mismatch issues that https://github.com/hashicorp/terraform-provider-aws/pull/16004 doesn't handle yet, and updates the docs, too!

I like this pull request a lot!  😺

**Review by kevincantu — APPROVED** [review]

File: aws/resource_aws_lb_target_group_test.go
Comment (fivehole68) [reply aws/resource_aws_lb_target_group_test.go]: Looks like that is correct I fixed it.

File: tools/go.mod
Comment (fivehole68) [reply tools/go.mod:10]: I think all of go.mod files were mistakes.  At the time it was required because I needed the latest aws-go lib.  I attempted to revert it but I don't know if I did it correctly.  If it isn't I would welcome some help.

Comment (kevincantu) [reply tools/go.mod:10]: Hmm, you've built a little merge-history puzzle!

Using an alias [I have defined](https://gist.github.com/kevincantu/aa5b6875a3fdebf71f1241b56f4f6375):
```
% git la -8          
*   6bd6a2037 <Mike> (HEAD, fivehole/f-lb-target-group-protocol-version) Merge branch 'f-lb-target-group-protocol-version' of https://github.com/fivehole68/terraform-provider-aws into f-lb-target-group-protocol-version [ae448c373 b5065c049]
|\  
| * b5065c049 <Mike> fix terrafmt errors [e93c7ae6b]
| *   e93c7ae6b <fivehole68> Merge branch 'master' into f-lb-target-group-protocol-version [c76539c8c f66074baa]
| |\  
| * | c76539c8c <Mike> Remove comment [484151258]
| * | 484151258 <Mike> Add support for protocol version to target groups [b1d2e1dbc]
* | | ae448c373 <Mike> Revert go files [74d6b7c9d]
* | | 74d6b7c9d <Mike> Add support for protocol version to target groups [f66074baa]
| |/  
|/|   
* | f66074baa <Dirk Avery> Update CHANGELOG.md [2713c9186]
```

I know I mentioned rebasing, but probably the easier next step is to just merge in the latest master and overwrite a couple of those go.mod and go.sum files.

Something like this might do the trick:
```
# your branch
git checkout f-lb-target-group-protocol-version
git log --decorate --graph --oneline -10
git diff --stat hashicorp/master f-lb-target-group-protocol-version

# fetch and merge in the latest upstream
git remote -v # below I assume Hashicorp's remote is named hashicorp
git fetch hashicorp
git merge hashicorp/master

# now look at your branch
git log --decorate --graph --oneline -10
git diff --stat hashicorp/master f-lb-target-group-protocol-version

# use checkout to overwrite a couple unintentional differences
git checkout hashicorp/master tools/go.mod
git checkout hashicorp/master tools/go.sum
git commit -am "Fix dep file merge issues"
...
```

Comment (fivehole68) [reply tools/go.mod]: Looks like that worked.  Thanks

**Review by kevincantu — APPROVED** [review]

**Comment by oliveirafilipe** [comment]
If you allow me, one advice that I could give to you is use `git rebase -i` to format your commits in a more consistent way before merging this PR.  

Hope this could be merge soon :rocket:

**Comment by kevincantu** [comment]
The maintainers could also just do a squish merge to clean it all up in a click: what do they prefer?

**Comment by fivehole68** [comment]
Is there anything else that is waiting on me to get this merged?

**Comment by kevincantu** [comment]
I've posted a comment here about this too:
https://discuss.hashicorp.com/t/can-a-maintainer-take-a-look-at-this-pull-req-to-add-support-for-aws-alb-protocol-versions/19161

**Comment by rpsadarangani** [comment]
Hi, is there any update on this PR when this is going to get merge.

**Comment by yliu-d** [comment]
@breathingdust ☝️ Can we have someone from Hashicorp to review this PR? Thanks!

**Comment by breathingdust** [comment]
Hi all 👋, just a quick update that this is in the queue for review and inclusion. Shouldn't be too much longer, we appreciate your patience!

**Comment by bill-rich** [comment]
Hi @fivehole68! Thank you for getting this enhancement together. I did some work on rebasing and making a few changes. Since I wasn't able to push to your branch, I opened a new PR (#17534). It still includes your commits, but will make it easier get it merged quickly. I'm going to close this PR so we can track things in the new one.

**Pull request closed without merging**
